### PR TITLE
ID-1276 Add azure cpu time to bard events

### PR DESCRIPTION
**Pull request by Ghost-in-a-Jar**
### Description

https://broadworkbench.atlassian.net/browse/ID-1276

This pr adds azure task cpu start and end time metrics for bard events. I referenced this merged pr to figure out how to get start and end time: https://github.com/broadinstitute/cromwell/pull/7415

terra helmfile pr:
https://github.com/broadinstitute/terra-helmfile/pull/5674

leo pr:
https://github.com/DataBiosphere/leonardo/pull/4664

#### `CHANGELOG.md`
 - [ ] I updated `CHANGELOG.md` in this PR
 - [x] I assert that this change shouldn't be included in `CHANGELOG.md` because it doesn't impact community users

#### Terra Release Notes
 - [ ] I added a suggested release notes entry in this Jira ticket
 - [x] I assert that this change doesn't need Jira release notes because it doesn't impact Terra users

### CODE REVIEW

File: supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala [hunk 1]
```diff
@@ -311,12 +318,31 @@ object TesAsyncBackendJobExecutionActor {
...
+      try {
+        val parsedStartTime = OffsetDateTime.parse(startTime)
+        val parsedEndTime = OffsetDateTime.parse(endTime)
+        Some(StartAndEndTimes(parsedStartTime, Option(parsedStartTime), parsedEndTime))
+      } catch {
+        case dateTimeParseException: DateTimeParseException =>
+          incrementFn(NonEmptyList.of("parse_tes_timestamp", "failure"), Some("bard"))
+          logger.error(s"Parsing TES task start and end time failed: $dateTimeParseException")
+          None
+      }
```
Comment (aednichols) [inline supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:336]: Would a `Try` and `toOption` work here instead of trying to make exceptions work with options?

Comment (Ghost-in-a-Jar) [reply supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:336]: Oh yeah good call!

Comment (Ghost-in-a-Jar) [reply supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:336]: I do want to have some sort of logging/metrics on parsing failures though so I think toOption wont work with that.

Comment (Ghost-in-a-Jar) [reply supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala]: I did a little refactor to use Try instead

File: supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala [hunk 2]
```diff
@@ -289,8 +294,9 @@ object TesAsyncBackendJobExecutionActor {
         val tesVmCostData = for {
           responseLogs <- t.logs
           startTime <- responseLogs.headOption.map(_.start_time)
+          endTime <- responseLogs.headOption.map(_.end_time)
```
Comment (THWiseman) [inline supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:297]: I think that this line _might_ cause issues. There's some somewhat involved polling logic happening. 

tl;dr I don't think we want to require that an `endTime` exist in order to establish `TesVmCostData`, as the presence of that object currently controls Cromwell's polling behavior

My understanding of the polling behavior:
- Cromwell submits a task to TES and polls every once in awhile.
- Cromwell tries to be efficient and not ask for more data than is necessary, and not ask for the cost data after it has been successfully obtained once. This  gets us cost data as soon as it's available, which we can then begin to show on the front end
- Because TES can't send us the end time until the task actually finishes, this change is going to make cromwell do the 'big poll' for the duration of the entire task. 

tl;dr - I don't think we want to require that the `endTime` be known for us to construct a `TesVmCostData`, as the presence of that object is how we control what we poll for.

Comment (THWiseman) [reply supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:297]: I wonder if we can accomplish everything desired for bard in the `onTaskComplete` function. We'll know the start time for sure, and we can either use the current time for the end time (not 100% accurate, but maybe close enough?) or query TES one last time to ask it for the end time.

File: supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala [hunk 3]
```diff
@@ -512,7 +536,16 @@ class TesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
...
     taskEndTime.onComplete {
-      case Success(result) => result.map(r => tellMetadata(Map(CallMetadataKeys.TaskEndTime -> r)))
+      case Success(result) =>
+        result.foreach(r => tellMetadata(Map(CallMetadataKeys.TaskEndTime -> r)))
+        val newCostData = runStatus.costData.map(_.copy(endTime = result))
+        runStatus match {
+          case _: Complete => tellBard(Complete(newCostData))
+          case _: Cancelled => tellBard(Cancelled(newCostData))
+          case failed: Failed => tellBard(Failed(sysLogs = failed.sysLogs, costData = newCostData))
+          case error: Error => tellBard(Error(sysLogs = error.sysLogs, costData = newCostData))
```
Comment (Ghost-in-a-Jar) [inline supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:546]: unfortunately you cant .copy a sealed trait so this seemed like the next best thing to me.

File: supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala [hunk 4]
```diff
@@ -290,7 +295,7 @@ object TesAsyncBackendJobExecutionActor {
           responseLogs <- t.logs
           startTime <- responseLogs.headOption.map(_.start_time)
           vmCost <- responseLogs.headOption.map(_.metadata.flatMap(_.get("vm_price_per_hour_usd")))
-          tesVmCostData = TesVmCostData(startTime, vmCost)
+          tesVmCostData = TesVmCostData(startTime, None, vmCost)
```
Comment (THWiseman) [inline supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:298]:
```suggestion
          // NB: End time is omitted here so we don't keep polling for it while the task runs. It will be acquired with a separate request when the task completes. 
          tesVmCostData = TesVmCostData(startTime, None, vmCost)
```

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 5]
```diff
@@ -1332,7 +1334,6 @@ trait StandardAsyncExecutionActor
         val metadata = getTerminalMetadata(state)
         onTaskComplete(state, oldHandle)
         tellMetadata(metadata)
-        tellBard(state)
```
Comment (Ghost-in-a-Jar) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala:1335]: moved this into onTaskComplete

File: supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala [hunk 6]
```diff
@@ -311,12 +317,71 @@ object TesAsyncBackendJobExecutionActor {
...
+    taskEndTime.onComplete {
+      case Success(result) =>
+        result.foreach(r => tellMetadataFn(Map(CallMetadataKeys.TaskEndTime -> r)))
+        val newCostData = runStatus.costData.map(_.copy(endTime = result))
+        runStatus match {
+          case _: Complete => tellBardFn(Complete(newCostData))
+          case _: Cancelled => tellBardFn(Cancelled(newCostData))
+          case failed: Failed => tellBardFn(Failed(sysLogs = failed.sysLogs, costData = newCostData))
+          case error: Error => tellBardFn(Error(sysLogs = error.sysLogs, costData = newCostData))
+          case _ => ()
```
Comment (Ghost-in-a-Jar) [inline supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala:355]: This onTaskComplete method is identical to before except for the new runStatus match and tellBard call

**Review by jgainerdewar — APPROVED** [review]
LGTM!

**Review by THWiseman — APPROVED** [review]
